storage/client_range_test.go

```diff
@@ -150,3 +154,156 @@ func TestRejectFutureCommand(t *testing.T) {
 		t.Errorf("expected 15, got %v", v)
 	}
 }
+
+// Test a case where a put operation of an older timestamp comes after
+// a put operation of a newer timestamp.
+//
+// 1) Writer executes a put operation with time T in a txn.
+// 2) Before the txn is committed, Reader sends a get
+//    operation with time T+100. This triggers the txn restart.
+// 3) Reader sends another get operation with time T+200. The
+//    write intent is resolved (and the txn timestamp is pushed to
+//    T+200), but the actual get operation has not yet been executed (and
+//    hence the timestamp cache has not been updated).
+// 4) Writer restarts the txn and executes the put operation
+//    again. The timestamp of the operation is T+100, and it will be
+//    ignored.
+func TestOutOfOrderPut(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	manualClock := hlc.NewManualClock(0)
+	clock := hlc.NewClock(manualClock.UnixNano)
+	store, stopper := createTestStoreWithEngine(t,
+		engine.NewInMem(proto.Attributes{}, 10<<20),
+		clock,
+		true,
+		nil)
+	defer stopper.Stop()
+
+	// Put an initial value.
+	key := "key"
+	initVal := []byte("initVal")
+	err := store.DB().Put(key, initVal)
+	if err != nil {
+		t.Fatalf("failed to get: %s", err)
+	}
+
+	waitPut := make(chan struct{})
+	waitFirstGet := make(chan struct{})
+	waitTxnRestart := make(chan struct{})
+	waitSecondGet := make(chan struct{})
+	waitTxnComplete := make(chan struct{})
+
+	// Start a writer.
+	go func() {
+		epoch := -1
+		// Start a txn that does read-after-write.
+		// The txn will be restarted twice, and the out-of-order put
+		// will happen in the second epoch.
+		if err := store.DB().Txn(func(txn *client.Txn) error {
+			epoch++
+
+			if epoch == 1 {
+				// Wait until the second get operation is issued.
+				close(waitTxnRestart)
+				<-waitSecondGet
+			}
+
+			updatedVal := []byte("updatedVal")
+			if err := txn.Put(key, updatedVal); err != nil {
+				return err
+			}
+
+			actual, err := txn.Get(key)
+			if err != nil {
+				return err
+			}
+			if !bytes.Equal(actual.ValueBytes(), updatedVal) {
+				t.Fatalf("Unexpected get result: %s", actual)
+			}
+
+			if epoch == 0 {
+				// Wait until the first get operation will push the txn timestamp.
+				close(waitPut)
+				<-waitFirstGet
+			}
+
+			b := &client.Batch{}
+			err = txn.Commit(b)
+			return err
+		}); err != nil {
+			t.Fatal(err)
+		}
+
+		if epoch != 2 {
+			t.Fatalf("Unexpected number of txn retry: %d", epoch)
+		}
+
+		close(waitTxnComplete)
+	}()
+
+	<-waitPut
+
+	// Advance the clock and send a get operation with higher
+	// priority to trigger the txn restart.
+	manualClock.Increment(100)
+
+	priority := int32(math.MaxInt32)
+	requestHeader := proto.RequestHeader{
+		Key:          proto.Key(key),
+		RaftID:       1,
+		Replica:      proto.Replica{StoreID: store.StoreID()},
+		UserPriority: &priority,
+		Timestamp:    clock.Now(),
+	}
+	getCall := proto.Call{
```

Comment:
Hmm, tried to use, but the code didn't look better since `proto.GetCall` returns a `Call` and then we need to update `Call.Args.RequestHeader.UserPriority`, etc.